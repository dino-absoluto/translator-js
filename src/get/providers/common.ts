/**
 * @file Common code
 * @author Dino <dinoabsoluto+dev@gmail.com>
 * @license
 * This file is part of translator-js.
 *
 * translator-js is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * translator-js is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with translator-js.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
/* imports */
/* code */

export type ContextCallback = (name: string) => Buffer | string[]
export abstract class Context {
  abstract requestFile (
    name: string,
    fn: ContextCallback): void

  parseNode (node: Node): string {
    return node.textContent || ''
  }
}

export type Content = (fmt: Context) => void

export interface ChapterData {
  group?: string
  name: string
  updateId?: string
  content?: Content
}

export interface Chapter extends ChapterData {
  fetch (): Promise<void>
}

export interface NovelData {
  readonly id: string
  name?: string
  author?: string
  description?: string
  keywords?: string[]
  genre?: string
  status?: {
    completed: boolean
    size: number
  }
}

export interface Novel extends NovelData {
  fetch (): Promise<void>
  fetchIndex (): Promise<Chapter[]>
}

export interface Provider {
  readonly acceptDomains: string[]
  fromURL (href: URL): Novel | undefined
}
