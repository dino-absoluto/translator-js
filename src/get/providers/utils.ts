/**
 * @author Dino <dinoabsoluto+dev@gmail.com>
 * @license
 * Copyright 2019 Dino <dinoabsoluto+dev@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
/* imports */
import { flow, trim } from '../../utils/flow'
import { JSDOM } from 'jsdom'
import * as mime from 'mime'
import fetch from '../../utils/syosetu-fetch'

/* code */
export const fetchDOM = async (url: string): Promise<Document> => {
  let { window: { document: doc } } =
    new JSDOM(await (await fetch(url)).text(), { url })
  return doc
}

interface FetchedFile {
  url: string
  buf: Buffer
  name: string
}

export const fetchFile = async (url: string): Promise<FetchedFile> => {
  const res = await fetch(url)
  const buf = await res.buffer()
  {
    const disp = res.headers.get('content-disposition')
    if (disp) {
      const name = trim(flow(disp.match(/; filename="[^"]+"/))
        .then((a): string => a[1]).value)
      if (name) {
        return {
          url,
          buf,
          name
        }
      }
    }
  }
  {
    const type = res.headers.get('content-type')
    if (type) {
      const ext = flow(trim(type.split(';')[0]))
        .then(mime.getExtension).value
      if (ext) {
        return {
          url,
          buf,
          name: `untitled.${ext}`
        }
      }
    }
  }
  return {
    url,
    buf,
    name: 'untitled'
  }
}
