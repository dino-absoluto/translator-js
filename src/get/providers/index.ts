/**
 * @file index.ts
 * @author Dino <dinoabsoluto+dev@gmail.com>
 * @license
 * This file is part of translator-js.
 *
 * translator-js is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * translator-js is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with translator-js.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
/* imports */
import { Provider, Novel } from './common'
/* code */
export { Provider, Novel, Chapter, Content, Formatter } from './common'

let initialized: Promise<void>
const domains = new Map()

export const register = (provider: Provider) => {
  for (const domain of provider.acceptDomains) {
    if (domains.has(domain)) {
      throw new Error(`'${domain}' domain existed`)
    }
    domains.set(domain, provider)
  }
}

export const getProvider = async (url: URL): Promise<Provider> => {
  if (!initialized) {
    initialized = (async () => {
      register((await import('./syosetu')).default)
    })()
    await initialized
    return getProvider(url)
  }
  let value = domains.get(url.hostname)
  if (!value) {
    throw new Error('No matching provider')
  }
  return value
}

export const getNovel = async (url: URL): Promise<Novel> => {
  const provider = await getProvider(url)
  let value = provider.fromURL(url)
  if (!value) {
    throw new Error('Failed to create Novel instance')
  }
  return value
}
