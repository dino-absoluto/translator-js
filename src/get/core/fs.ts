/**
 * @author Dino <dinoabsoluto+dev@gmail.com>
 * @license
 * Copyright 2019 Dino <dinoabsoluto+dev@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
/* imports */
import { trim, flow } from '../../utils/flow'
import * as pfs from '../../utils/pfs'
import * as path from 'path'
import makeDir = require('make-dir')
/* code */

const sanitizeName = (name?: string): string | undefined => {
  return flow(trim(name))
    .then(path.normalize)
    .then((name): string | undefined => {
      if (path.basename(name) !== name) {
        return
      }
      if (name.startsWith('..')) {
        return
      }
      return name
    }).value
}

interface FSItem {
  parent?: Folder
  path: string
  real (): Promise<void>
  remove (): Promise<void>
  close (): Promise<void>
  rename (name: string): Promise<void>
}

export class File implements FSItem {
  public readonly parent: Folder
  private _filename?: string

  public constructor (parent: Folder, name: string) {
    this.parent = parent
    this._filename = sanitizeName(name)
  }

  public get path (): string {
    const { parent, _filename } = this
    if (!_filename) {
      throw new Error('filename is not set')
    }
    return path.join(parent.path, _filename)
  }

  public async real (): Promise<void> {
    if (!this._filename) {
      throw new Error('filename is not set')
    }
  }

  public async rename (name: string): Promise<void> {
    const newName = sanitizeName(name)
    if (!newName) {
      throw new Error('filename is invalid')
    }
    const oldName = this._filename
    this._filename = newName
    if (oldName) {
      if (newName === oldName) {
        return
      }
      try {
        const parentDir = this.parent.path
        await pfs.rename(
          path.join(parentDir, oldName),
          path.join(parentDir, newName))
      } catch (err) {
        if (err.code !== 'ENOENT') {
          throw err
        }
      }
    }
  }

  public async close (): Promise<void> {
    delete this._filename
    this.parent.children.delete(this)
  }

  public async access (mode?: number): Promise<void> {
    return pfs.access(this.path, mode)
  }

  public async remove (): Promise<void> {
    await this.real()
    const fpath = this.path
    await this.close()
    try {
      await pfs.unlink(fpath)
    } catch (err) {
      if (err.code !== 'ENOENT') {
        throw err
      }
    }
  }

  public async read (): Promise<Buffer> {
    await this.real()
    return pfs.readFile(this.path)
  }

  public async write (content: string | Buffer): Promise<void> {
    await this.parent.real()
    await this.real()
    await pfs.writeFile(this.path, content)
  }
}

export class Folder implements FSItem {
  public readonly children: Set<Folder | File> = new Set()
  public readonly parent?: Folder
  /* private data */
  private _dirname?: string
  private _accessed = false

  public constructor (parent: Folder | undefined | null, name: string) {
    if (parent) {
      this.parent = parent
      this._dirname = sanitizeName(name)
    } else {
      this._dirname = path.resolve(name)
    }
  }

  public get name (): string | undefined { return this._dirname }
  public get path (): string {
    const { parent, _dirname } = this
    if (!_dirname) {
      throw new Error('Folder\'s path is undefined')
    }
    if (parent) {
      return path.join(parent.path, _dirname)
    }
    return _dirname
  }

  public async real (): Promise<void> {
    if (this._accessed) {
    } else {
      await makeDir(this.path)
      this._accessed = true
    }
  }

  public async close (): Promise<void> {
    delete this._dirname
    if (this.parent) {
      this.parent.children.delete(this)
    }
  }

  public async remove (): Promise<void> {
    const { children } = this
    if (!children.size) {
      throw new Error('folder is not empty')
    }
    const fpath = this.path
    await this.close()
    await pfs.rmdir(fpath)
  }

  public get renameable (): boolean { return !!this.parent }
  public async rename (name: string): Promise<void> {
    if (!this.parent) {
      throw new Error('folder cannot be renamed')
    }
    const newName = sanitizeName(name)
    if (!newName) {
      throw new Error('folder name is invalid')
    }
    const oldName = this._dirname
    this._dirname = newName
    this._accessed = false
    if (oldName) {
      if (newName === oldName) {
        return
      }
      try {
        const parentDir = this.parent.path
        await pfs.rename(
          path.join(parentDir, oldName),
          path.join(parentDir, newName))
      } catch (err) {
        if (err.code !== 'ENOENT') {
          throw err
        }
      }
    }
  }

  public requestFolder (name: string): Folder {
    const entry = new Folder(this, name)
    this.children.add(entry)
    return entry
  }

  public requestFile (name: string): File {
    const entry = new File(this, name)
    this.children.add(entry)
    return entry
  }
}
