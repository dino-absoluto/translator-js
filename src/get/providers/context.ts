/**
 * @author Dino <dinoabsoluto+dev@gmail.com>
 * @license
 * Copyright 2019 Dino <dinoabsoluto+dev@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
/* imports */
import { flow, trim } from '../../utils/flow'
import { JSDOM } from 'jsdom'
import * as path from 'path'
import filenamify = require('filenamify')
/* code */

/* eslint-disable-next-line @typescript-eslint/explicit-function-return-type */
const Node = flow(new JSDOM()).then(dom => {
  return dom.window.Node
}).value

interface TokenBr {
  type: 'br'
}

interface TokenString {
  type: 'text'
  text: string
}

interface TokenRuby {
  type: 'ruby'
  text: string
  ruby: string
}

interface TokenImage {
  type: 'image'
  url: string
  text: string
}

interface TokenLink {
  type: 'link'
  url: string
  text: string
}

export type Token = TokenBr | TokenString | TokenRuby | TokenImage | TokenLink

const isElementNode = (node: Node): node is HTMLElement =>
  node.nodeType === Node.ELEMENT_NODE

export type ContextCallback = (name: string) => Buffer | string[]

export abstract class Context {
  private names = new Set<string>()
  private urls = new Map<string, string>()
  public abstract requestFile (
    name: string,
    fn: ContextCallback): void
  public resolveName (name: string): string | undefined {
    name = path.normalize(trim(name) || '')
    if (name === '.') {
      return
    }
    const ext = path.extname(name)
    const base = filenamify(path.basename(name, ext))
    name = base + ext
    const { names } = this
    if (!names.has(name)) {
      names.add(name)
      return name
    }
    for (let i = 1; i < 100; i++) {
      const name = `${base} (${i})${ext}`
      if (!names.has(name)) {
        names.add(name)
        return name
      }
    }
    return ''
  }

  public mapURL (url: string, filename: string): void {
    this.urls.set(url, filename)
  }

  public resolveURL (url: string): string {
    return this.urls.get(url) || url
  }

  public parseNode (node: Node): string {
    return this.render(Context.tokenize(node))
  }

  public render (tokens: Token[]): string {
    let text = ''
    for (const tok of tokens) {
      let tt: string
      switch (tok.type) {
        case 'text': {
          tt = tok.text
          break
        }
        case 'ruby': {
          tt = `${tok.text}(${tok.ruby})`
          break
        }
        case 'link': {
          tt = `[${tok.text}](${this.resolveURL(tok.url)})`
          break
        }
        case 'image': {
          tt = `![${tok.text}](${this.resolveURL(tok.url)})`
          break
        }
        case 'br': {
          tt = '\n'
          break
        }
        default:
          tt = ''
          break
      }
      text += tt
    }
    return text
  }

  public static tokenizeArray (nodes: NodeList): Token[] {
    return [...nodes].reduce((acc, cnode): Token[] => {
      return acc.concat(Context.tokenize(cnode))
    }, [] as Token[])
  }

  public static tokenize (node: Node): Token[] {
    if (node.nodeType === Node.TEXT_NODE) {
      return [{
        type: 'text',
        text: (node.textContent || '').replace(/^\n|\n$/g, '')
      }]
    }
    if (!isElementNode(node)) {
      return []
    }
    switch (node.nodeName) {
      case 'RUBY': {
        const text = trim(flow(node.querySelector('rb'))
          .then((rb): string | null => rb.textContent)
          .value ||
        flow([...node.children]).then((children): string => {
          let text = ''
          for (const node of children) {
            if (node.nodeType === Node.TEXT_NODE) {
              text += node.textContent
            }
          }
          return text
        }).value || node.textContent)
        if (text) {
          const ruby = flow(node.querySelector('rt')).then(
            (rt): string | undefined => {
              return trim(rt.textContent)
            }).value
          if (ruby && text !== ruby) {
            return [{
              type: 'ruby',
              text,
              ruby
            }]
          } else {
            return [{
              type: 'text',
              text
            }]
          }
        } else {
          return []
        }
      }
      case 'BR': {
        if (flow(node.nextSibling)
          .then((n): string | undefined =>
            (n.nodeType === Node.TEXT_NODE && n.textContent) || undefined)
          .then((t): boolean => t.startsWith('\n'))) {
          return []
        }
        return [{
          type: 'br'
        }]
      }
      case 'A': {
        const a = node as HTMLLinkElement
        const tok: Token = {
          type: 'link',
          text: a.textContent || a.href,
          url: a.href
        }
        const tokens = [...a.querySelectorAll('img')].reduce(
          (all, node): Token[] => all.concat(this.tokenize(node)), [] as Token[])
        if (tokens.length) {
          tokens.push({
            type: 'br'
          })
          tokens.push(tok)
          return tokens
        }
        return [tok]
      }
      case 'IMG': {
        const img = node as HTMLImageElement
        return [{
          type: 'image',
          text: img.alt,
          url: img.src
        }]
      }
      case 'IFRAME':
      case 'VIDEO':
      case 'EMBEDED': {
        return []
      }
      case 'P': {
        const tokens = Context.tokenizeArray(node.childNodes)
        tokens.push({
          type: 'br'
        })
        return tokens
      }
      default: {
        return Context.tokenizeArray(node.childNodes)
      }
    }
    // return []
  }
}

export class SimpleContext extends Context {
  public text: ([string, string[]])[] = []
  public bufs: ([string, Buffer])[] = []
  public requestFile (unsafeName: string, get: ContextCallback): void {
    const name = this.resolveName(unsafeName)
    if (!name) {
      return
    }
    let data = get(name)
    if (Array.isArray(data)) {
      this.text.push([name, data as string[]])
    } else {
      this.bufs.push([name, data as Buffer])
    }
  }
}
