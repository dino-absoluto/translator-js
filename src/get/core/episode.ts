/**
 * @author Dino <dinoabsoluto+dev@gmail.com>
 * @license
 * Copyright 2019 Dino <dinoabsoluto+dev@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
/* imports */
import { flow } from '../../utils/flow'
import { Chapter } from '../providers/common'
import { Context as AbstractContext } from '../providers/context'
import { Folder, File } from './fs'
import { gunzipSync, gzipSync } from 'zlib'
import pLimit from 'p-limit'
import throttle = require('lodash/throttle')
/* code */

export type ProgressFn = (done: number, total: number) => void
export interface UpdateOptions {
  progress?: ProgressFn
  checkFs?: boolean
}

export interface UpdateItem {
  index: number
  chapter: Chapter
}
export interface UpdateReport {
  updates: UpdateItem[]
  news: UpdateItem[]
}

class Context extends AbstractContext {
  readonly files = new Map<string, string[] | Buffer>()
  readonly prefix: string
  constructor (prefix: string) {
    super()
    this.prefix = prefix
  }

  requestFile (name: string, get: (name: string) => string[] | Buffer) {
    const { files, prefix } = this
    flow(this.resolveName(`${prefix}${name}`.trim()))
      .then(name => {
        if (files.has(name)) {
          return
        }
        files.set(name, get(name))
      })
  }
}

interface EpisodeData {
  groupId?: number
  updateId?: string
  files?: string[]
}

interface EpisodeListData {
  groups: string[]
  episodes: EpisodeData[]
}

export class EpisodeList {
  readonly rootDir: Folder
  data: EpisodeListData
  ready: Promise<void>
  private folders: Folder[]
  private metaFile: File
  private compressedCache: boolean
  private pad0: number = 3
  private saveThrottle = throttle(this.save.bind(this), 3000)
  constructor (rootDir: Folder, options: {
    compressedCache?: boolean
    pad0?: number
    data?: EpisodeListData
  } = {}) {
    this.rootDir = rootDir
    this.data = options.data || {
      groups: [],
      episodes: []
    }
    this.compressedCache = !!options.compressedCache
    this.pad0 = options.pad0 || this.pad0
    const metaDir = rootDir.requestFolder('!meta')
    const metaFile = metaDir.requestFile(
      this.compressedCache ? '!cache.json.gz' : '!cache.json')
    const folders = [
      rootDir.requestFolder('0'.padStart(this.pad0, '0'))
    ]
    this.folders = folders
    this.metaFile = metaFile
    this.ready = this.load()
  }

  async updateWith (newData: Chapter[], options: UpdateOptions = {}) {
    const { data: { episodes } } = this
    const { groups, chapters } = this.preprocess(newData)
    const { progress } = options
    const report: UpdateReport = {
      updates: [],
      news: []
    }
    const limit = pLimit(2)
    const oldLength = episodes.length
    await this.updateGroups(groups)
    const tasks: UpdateItem[] = []
    for (const [index, chapter] of chapters.entries()) {
      if (await this.pingEpisode(index, chapter, options)) {
        const item: UpdateItem = {
          index, chapter
        }
        tasks.push(item)
        if (index < oldLength) {
          report.updates.push(item)
        } else {
          report.news.push(item)
        }
      }
    }
    let count = 0
    const length = tasks.length
    if (progress) {
      progress(0, length)
    }
    await Promise.all(tasks.map(({ index, chapter }) => {
      return limit(async () => {
        await this.updateEpisode(index, chapter, options)
        await this.saveThrottle()
        if (progress) {
          progress(++count, length)
        }
      })
    }))
    episodes.length = chapters.length
    await this.saveThrottle.flush()
    return report
  }

  private encode (data: object) {
    if (!this.compressedCache) {
      return JSON.stringify(data, null, 1)
    } else {
      return gzipSync(JSON.stringify(data))
    }
  }
  private decode (data: Buffer) {
    if (!this.compressedCache) {
      return JSON.parse(data.toString())
    } else {
      return JSON.parse(gunzipSync(data).toString())
    }
  }

  private async updateGroups (groups: string[]) {
    const { folders, rootDir, data } = this
    for (const [index, group] of groups.entries()) {
      const id = index + 1
      let folder = folders[id]
      if (!folder) {
        folder = rootDir.requestFolder(group)
        folders[id] = folder
      } else {
        if (group !== folder.name) {
          await folder.rename(group)
        }
      }
    }
    data.groups = groups
  }

  private async isEpisodeUptodate (
    ep: EpisodeData,
    ch: Chapter & EpisodeData,
    options: UpdateOptions
  ): Promise<boolean> {
    if (!ep.files || !ep.files.length) {
      return false
    }
    if (ch.updateId && ep.updateId !== ch.updateId) {
      return false
    }
    if (ch.groupId && ep.groupId !== ch.groupId) {
      return false
    }
    const folder = this.folders[ep.groupId || 0]
    if (!folder) {
      throw new Error(`Folder must not be ${folder}`)
    }
    if (!options.checkFs) {
      return true
    }
    try {
      for (const fpath of ep.files) {
        const file = folder.requestFile(fpath)
        await file.access()
        await file.close()
      }
      return true
    } catch {
      return false
    }
  }

  private async pingEpisode (
    index: number,
    ch: Chapter & EpisodeData,
    options: UpdateOptions
  ) {
    const { data: { episodes } } = this
    let ep = episodes[index]
    if (!ep) {
      ep = {}
      episodes[index] = ep
    }
    if (await this.isEpisodeUptodate(ep, ch, options)) {
      return false
    }
    return ep
  }

  private async updateEpisode (
    index: number,
    ch: Chapter & EpisodeData,
    options: UpdateOptions
  ) {
    const ep = await this.pingEpisode(index, ch, options)
    if (!ep) {
      return
    }
    if (ep.files) {
      const folder = this.folders[ep.groupId || 0]
      // remove old files
      await Promise.all(ep.files.map(file => folder.requestFile(file).remove()))
      delete ep.files
    }
    await ch.fetch()
    ep.groupId = ch.groupId
    ep.updateId = ch.updateId
    const folder = this.folders[ep.groupId || 0]
    if (!folder) {
      throw new Error(`Folder must not be ${folder}`)
    }
    const context = new Context(
      `${(index + 1).toString().padStart(this.pad0, '0')} `)
    if (ch.content) {
      ch.content(context)
      delete ch.content
    }
    // write files
    const files = []
    for (const [name, data] of context.files) {
      let buf: Buffer
      if (data instanceof Buffer) {
        buf = data
      } else {
        buf = flow(data.join('\n\n---\n\n'))
          .then(text => {
            if (text[text.length - 1] !== '\n') {
              return text + '\n'
            } else {
              return text
            }
          }).then(Buffer.from).value
      }
      files.push(name)
      const file = folder.requestFile(name)
      await file.write(buf)
      await file.close()
    }
    if (files.length) {
      ep.files = files
    }
  }

  private preprocess (chapters: Chapter[]) {
    let groupId = 0
    let groupName: string | undefined
    const groups = []
    let count = 1
    for (const chapter of chapters) {
      const ch: Chapter & EpisodeData = chapter
      ++count
      if (ch.group !== groupName || count > 50) {
        count = 1
        groupId = groups.length + 1
        groups.push(`${
          groupId.toString().padStart(this.pad0, '0')} ${
          ch.group || ''}`.trim())
        groupName = ch.group
      }
      ch.groupId = groupId
    }
    return {
      groups,
      chapters
    }
  }

  private async load () {
    const { metaFile } = this
    try {
      const cache = await this.decode(await metaFile.read())
      const { data } = this
      await this.updateGroups(cache.groups)
      data.episodes = cache.episodes || []
    } catch (err) {
      if (err.code !== 'ENOENT') {
        throw err
      }
    }
  }

  async save () {
    const { metaFile } = this
    await metaFile.write(await this.encode(this.data))
  }
}
