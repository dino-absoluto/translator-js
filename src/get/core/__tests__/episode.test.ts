/**
 * @author Dino <dinoabsoluto+dev@gmail.com>
 * @license
 * Copyright 2019 Dino <dinoabsoluto+dev@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
/* imports */
import { EpisodeList } from '../episode'
import { Folder } from '../fs'
import { Chapter, ChapterData, RenderFn } from '../../providers/common'
import { hashDir, setup } from '../../../utils/test-utils'
import * as path from 'path'

/* setup */
const { __TMPDIR } = setup(__filename)

/* code */
interface FakeData extends ChapterData {
  group?: string | undefined
  name: string
  updateId?: string | undefined
  text: string
}

class FakeChapter implements Chapter {
  public group?: string | undefined
  public name: string = ''
  public updateId?: string | undefined
  public content?: RenderFn
  private text: string = ''
  public constructor (options: FakeData) {
    Object.assign(this, options)
  }
  public async fetch (): Promise<void> {
    this.content = (fmt) => {
      fmt.requestFile(this.name + '.txt', _fname => [
        _fname,
        this.text
      ])
    }
  }
}

describe('EpisodeList', () => {
  test('constructor.1', async () => {
    const tmpPath = path.join(__TMPDIR, 'constructor')
    const rootDir = new Folder(null, tmpPath)
    const eplist = new EpisodeList(rootDir, {
    })
    await eplist.ready
  })
  test('simple tree', async () => {
    const tmpPath = path.join(__TMPDIR, 'constructor')
    const rootDir = new Folder(null, tmpPath)
    const eplist = new EpisodeList(rootDir, {
    })
    await rootDir.real()
    await eplist.ready
    expect(await hashDir(path.join(tmpPath, '**/*'), {
      cwd: tmpPath
    })).toMatchSnapshot()
    await eplist.updateWith([
      new FakeChapter({
        name: 'prologue',
        text: 'Hello!!'
      }),
      new FakeChapter({
        name: 'First Chapter',
        text: 'Nothing!!'
      }),
      new FakeChapter({
        name: '2nd Chapter',
        text: 'Still nothing!!'
      }),
      new FakeChapter({
        group: 'New group',
        name: '3nd Chapter',
        text: 'Still nothing!!!'
      })
    ])
    expect(await hashDir(path.join(tmpPath, '**/*'), {
      cwd: tmpPath
    })).toMatchSnapshot()
    await eplist.updateWith([
      new FakeChapter({
        name: 'prologue',
        text: 'Hello!! Updated!'
      }),
      new FakeChapter({
        updateId: '2',
        name: 'First Chapter',
        text: 'Nothing!! 2!!'
      }),
      new FakeChapter({
        updateId: '2',
        name: '2nd Chapter',
        text: 'Still nothing!! Again!'
      }),
      new FakeChapter({
        group: 'New group',
        name: '3nd Chapter',
        text: 'Still nothing!!! And Again!'
      })
    ])
    expect(await hashDir(path.join(tmpPath, '**/*'), {
      cwd: tmpPath
    })).toMatchSnapshot()
    await eplist.updateWith([
      new FakeChapter({
        name: 'prologue',
        text: 'Hello!! Updated!'
      }),
      new FakeChapter({
        updateId: '2',
        name: 'First Chapter',
        text: 'Nothing!! 2!!'
      }),
      new FakeChapter({
        group: 'New group',
        updateId: '2',
        name: '2nd Chapter',
        text: 'Still nothing!! Again!'
      }),
      new FakeChapter({
        group: 'New group',
        name: '3nd Chapter',
        text: 'Still nothing!!! And Again!'
      })
    ])
    expect(await hashDir(path.join(tmpPath, '**/*'), {
      cwd: tmpPath
    })).toMatchSnapshot()
  })
})
