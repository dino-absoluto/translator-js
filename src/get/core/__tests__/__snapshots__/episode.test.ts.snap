// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodeList simple tree 1`] = `Object {}`;

exports[`EpisodeList simple tree 2`] = `
Object {
  "!meta/!cache.json": "cI57C6vM9eBN8jM6HUNqKpmXIPzMfOPILdo6cIPqvF1djB+i4Wu8G3qnzqUBEGg/2NtDfwYT/byRswADNJmBwg==",
  "!meta/0 prologue.txt": "3d1Ihjf8wO68sy6cPHdwk60Ew2IWD/yHYKg4t6BHEwKTT9CZb61P1Y2Qd17j0NuhxP5YgzYUn9ezvVCxCKIYlQ==",
  "!meta/1 First Chapter.txt": "1gnXI8XrAp5oK0RSKLnKbyy1SsrjhsptZKgWQUv0lVW/IpKWItpz0sI6f367VojVjgwVMZvpbbjyv0uhrlOyOQ==",
  "!meta/2 2nd Chapter.txt": "1zPfLroO9Djx5U5Ru7qynA+iJd/X03YV86AAh0jKXNSrB5frmBkIWiRRg0G8nrBz2FUbHu7gKPp0BQPGKTzqIA==",
  "New group/3 3nd Chapter.txt": "S1hnudiG1WyOutI2+XbnIUGBZkGkoQ/lyQ+I3h7yqXv1V27dPgE5E8Zul8XUUZO8+LZ802Eyj1UZUrBtDz6qOw==",
}
`;

exports[`EpisodeList simple tree 3`] = `
Object {
  "!meta/!cache.json": "sImOQ7CmLvhaizsKsXqC3oEpc4fGjo5fdtKHtb7RKexiyKBuHBQdZdtchGHZti6Jj+wgrdtnwm+3LNJ7fQ/j/w==",
  "!meta/0 prologue.txt": "3d1Ihjf8wO68sy6cPHdwk60Ew2IWD/yHYKg4t6BHEwKTT9CZb61P1Y2Qd17j0NuhxP5YgzYUn9ezvVCxCKIYlQ==",
  "!meta/1 First Chapter.txt": "WGerlyTOlfhqvFLg8WdfymXNNAOAZI1Uf4epXC9RiMYuPmS1sApcHHbgPCS8PyGRk6sjnQ8qT3azkGFKm9QPiQ==",
  "!meta/2 2nd Chapter.txt": "86ZBM8TpNlz8rK6Oz1JzJNI+ScJ3ExU812QAe+Sig9kQJjH/lOVYxD/JvL5K66tEysoHviHgW/QHIIblo12vNA==",
  "New group/3 3nd Chapter.txt": "S1hnudiG1WyOutI2+XbnIUGBZkGkoQ/lyQ+I3h7yqXv1V27dPgE5E8Zul8XUUZO8+LZ802Eyj1UZUrBtDz6qOw==",
}
`;

exports[`EpisodeList simple tree 4`] = `
Object {
  "!meta/!cache.json": "RRW0uaY/aIilJYz9mRdN56Qdf797rDE1c6pNDvDRmO5P8GHaYxxcqsc9GInWALkcMTn87s+BsgGjb1k5SURfHw==",
  "!meta/0 prologue.txt": "3d1Ihjf8wO68sy6cPHdwk60Ew2IWD/yHYKg4t6BHEwKTT9CZb61P1Y2Qd17j0NuhxP5YgzYUn9ezvVCxCKIYlQ==",
  "!meta/1 First Chapter.txt": "WGerlyTOlfhqvFLg8WdfymXNNAOAZI1Uf4epXC9RiMYuPmS1sApcHHbgPCS8PyGRk6sjnQ8qT3azkGFKm9QPiQ==",
  "New group/2 2nd Chapter.txt": "86ZBM8TpNlz8rK6Oz1JzJNI+ScJ3ExU812QAe+Sig9kQJjH/lOVYxD/JvL5K66tEysoHviHgW/QHIIblo12vNA==",
  "New group/3 3nd Chapter.txt": "S1hnudiG1WyOutI2+XbnIUGBZkGkoQ/lyQ+I3h7yqXv1V27dPgE5E8Zul8XUUZO8+LZ802Eyj1UZUrBtDz6qOw==",
}
`;
